<a href="https://colab.research.google.com/github/piyusdev2006/AgenticAIHindi/blob/main/M1_Assignment_Stock_Price_Prediction_with_LLM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Module 1 No Code/Low Code Assignment: Stock Price Prediction with LLM (Using Gemini in Colab)

This assignment explores how large language models (LLMs) can perform complex tasks that previously required coding. Instead of manually writing code, we will interact with Google's Gemini (Click on the Gemini icon on the top right corner) in Google Colab to:

- Load and analyze stock price data.
- Engineer relevant features for prediction.
- Build and evaluate models such as regression and random forest.
- Discuss potential trading strategies based on predictions.

**Note**: While this notebook contains example code, students are encouraged to use LLMs to generate code snippets and explanations, managing interactions within the LLM's context window.

## Task Breakdown

### Step 1: Data Exploration
- **Prompt the LLM**: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance.
- **Key Considerations**: Ensure the model retains context while summarizing key statistics.

### Step 2: Feature Engineering
- **Prompt the LLM**: Generate meaningful features such as moving averages, technical indicators, and volatility measures.
- **Managing Context**: If the LLM forgets prior outputs, remind it by reintroducing key statistics.

### Step 3: Model Building
- **Prompt the LLM**: Train a model (regression for price prediction, random forest for trend classification).
- **Split Ratio Consideration**: Explain the importance of train-test split (e.g., 70/30) and verify the LLM respects it.

### Step 4: Model Evaluation & Prediction
- **Prompt the LLM**: Use the trained model to predict stock prices for 2023-2024.
- **Analyze Performance**: Interpret accuracy, R² scores, and classification reports.

### Step 5: Strategy Discussion
- **Prompt the LLM**: Based on model performance, discuss possible trading strategies.
- **Critical Thinking**: Evaluate whether the model is truly useful for financial decision-making.

---

Throughout the assignment, keep in mind that LLMs have a limited context window. If necessary, break down tasks into smaller chunks and reintroduce critical information.



## Using Gemini in Colab

To use Gemini directly in your browser within Google Colab:

1. **Create a new code cell** by clicking the `+ Code` button.
2. In the newly created cell, you will see a placeholder text: **"Start coding or generate with AI"**.
3. Click on **"Generate"** and enter your prompt, such as:
   > "Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics."
4. Gemini will generate code based on your prompt. You can then review, modify, and run it within the notebook.
5. If the model loses context, you may need to remind it of previous responses or break down tasks into smaller steps.

This method allows you to interactively generate and refine code without manually writing every line.


In [ ]:
# prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics.

!pip install yfinance

import yfinance as yf
import pandas as pd

# Download historical data for AAPL
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Display summary statistics
print(data.describe())


In [ ]:
# prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics.

!pip install yfinance

import yfinance as yf
import pandas as pd

# Download historical data for AAPL from 2020-01-01 to 2023-01-01
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Summarize key statistics
print(data.describe())


YF.download() has changed argument auto_adjust default to True


[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.800474  129.350921  126.136626  127.722421  1.120920e+08
std      30.461355   30.739156   30.179073   30.489492  5.602586e+07
min      54.449894   55.452160   51.595990   55.350235  3.519590e+07
25%     112.857315  114.354702  111.397715  113.037711  7.636470e+07
50%     133.185188  134.319768  131.222483  133.006499  9.493580e+07
75%     148.415882  149.584513  146.468716  147.782312  1.296327e+08
max     178.879913  179.793905  176.039592  179.489238  4.265100e+08


In [ ]:
# Prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01
# using Yahoo Finance and summarize its key statistics.

import yfinance as yf

data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Summarize key statistics
print(data.describe())



YF.download() has changed argument auto_adjust default to True


[*********************100%***********************]  1 of 1 completed


Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.800473  129.350921  126.136626  127.722421  1.120920e+08
std      30.461355   30.739156   30.179073   30.489492  5.602586e+07
min      54.449894   55.452160   51.595990   55.350235  3.519590e+07
25%     112.857307  114.354706  111.397708  113.037705  7.636470e+07
50%     133.185196  134.319791  131.222513  133.006484  9.493580e+07
75%     148.415874  149.584494  146.468701  147.782328  1.296327e+08
max     178.879898  179.793936  176.039622  179.489269  4.265100e+08


## Now try doing step 2 - step 5 by yourself

The sample prompts and LLM responses are listed below for your reference.

# Sample Prompts and LLM Responses
---

## Step 1: Data Exploration

### Prompt:
> Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 from Yahoo Finance, keep it in memory and summarize its key statistics.

### LLM Response:
- **Highest Closing Price:** \$182.01 on January 3, 2022
- **Lowest Closing Price:** \$56.01 on March 23, 2020
- **Average Closing Price:** \$126.69
- **Total Trading Days:** 756

---

## Step 2: Feature Engineering

### Prompt:
> Suggest some features that might help improve a stock price prediction model.

### LLM Response:
1. **Historical Prices**: Include past closing, opening, high, and low prices.
2. **Volume**: Daily trading volumes.
3. **Moving Averages**: Simple and exponential moving averages (e.g., 5-day, 10-day, 50-day).
4. **Technical Indicators**: RSI, MACD, Bollinger Bands, etc.
5. **Volatility**: Historical volatility, Average True Range (ATR).
6. **Sentiment Analysis**: News articles, social media sentiment.
7. **Economic Indicators**: Interest rates, inflation rates, GDP growth.
8. **Time-based Features**: Day of the week, month, year.
9. **Sector Performance**: Comparison with industry and sector indices.
10. **External Events**: Major geopolitical events, earnings reports.

---

## Step 3.1: Model Building - Linear Regression

### Prompt:
> Adopt the first four kinds of features and train a regression model to predict tomorrow's stock price. Split your training and testing data by a 70/30 ratio and report your performance metric.

### LLM Response:
- **Mean Squared Error (MSE):** 859.04
- **R-squared (R²) Score:** -0.009

These metrics indicate that the model is not performing well, as an R² score close to zero or negative suggests that the model does not explain the variability in the target variable better than a simple mean of the target variable. This outcome suggests that the chosen features and model might need adjustments or improvements to enhance prediction accuracy. ​​

---
(It might be too hard to predict the exact price but how about ups and downs?)
# Step 3.2: Model Building - Random Forest Classification

### Prompt:
> Create a binary target variable based on historical prices (1 if the close price increase tomorrow, 0 other), and train a Random Forest model using the same train/test split ratio. Report feature importance and performance metrics.

### LLM Response:
- **Accuracy:** 71.82%
- **Confusion Matrix:** [[84, 26],
 [36, 74]]
- **Classification Report:**

              precision    recall  f1-score   support

         0       0.70      0.76      0.73       110
         1       0.74      0.67      0.70       110
        accuracy                     0.72       220
        macro avg    0.72      0.72      0.72       220
        weighted avg  0.72      0.72      0.72       220



- **Feature Importance:** The most important feature for predicting whether the price will increase or decrease the next day is the closing price, followed by the 5-day moving average (MA_5). The model achieved an accuracy of approximately 71.82%, indicating a reasonable performance for this classification task.
         Feature	Importance
         Open	   0.0900
         High	   0.0883
         Low	   0.0980
         Close	   0.3087
         Volume	    0.0982
         MA_5	   0.1242
         MA_10	   0.0974
         MA_50	   0.0951

---

## Step 4: Model Evaluation & Prediction
### Prompt:
> Use this model to predict stock price increase/decrease for period starting from 2023-01-01 to 2024-01-01. Report your performance metrics.

### LLM Response:
The Random Forest classifier's performance on the new test data from 2023-01-01 to 2024-01-01 is as follows:
- **Accuracy:** 60.85%
- **Confusion Matrix:** [[105,   0],
 [ 83,  24]]
- **Classification Report:**

            precision    recall  f1-score   support

       0       0.56      1.00      0.72       105
       1       1.00      0.22      0.37       107

        accuracy                   0.61       212
        macro avg  0.78     0.61       0.54       212
        weighted avg 0.78    0.61        0.54       212

---
## Step 5: Strategy Discussion
### Prompt:
> What do you think about the performance of the model and what trading strategies could I build with this model?

### LLM Response:
- The model shows a moderate performance with an accuracy of approximately 60.85%. However, the precision and recall metrics indicate that while the model is good at identifying days when the stock price does not increase (class 0), it struggles significantly to identify days when the stock price does increase (class 1). This suggests a need for further feature engineering or model tuning to improve the prediction performance for this specific classification task.
- Potential Trading Strategies
Given the model's performance, potential trading strategies could be formulated as follows:

#### Conservative Strategy:

- **Focus**: Only trade on days when the model predicts a price increase with high confidence.
- **Action**: Buy stocks when the model predicts an increase (class 1).
- **Risk Management**: Since the model has a high precision for the no-increase class (class 0), it might be more reliable to stay out of the market when the model predicts no increase.

#### Contrarian Strategy:

- **Focus**: Take advantage of the model's high recall for no price increase predictions (class 0).
- **Action**: Sell short or avoid buying stocks on days when the model predicts no increase.
- **Risk Management**: This strategy assumes that the model’s predictions of no price increase are more reliable. It requires a careful assessment of market conditions to avoid false signals.

#### Hybrid Strategy:

- **Combine**: Use both conservative and contrarian strategies based on confidence thresholds.
- **Action**: Set a higher threshold for buying (only buy when the predicted probability of price increase is significantly high) and a lower threshold for selling or staying out of the market.
- **Risk Management**: Adjust thresholds dynamically based on recent model performance and market volatility.


---
## Conclusion

Through this assignment, we have explored how LLMs can assist in complex machine learning tasks without requiring extensive coding knowledge. Key takeaways:

- **Context Management**: LLMs have limited memory, so breaking tasks into smaller steps and reinforcing context is crucial.
- **Feature Engineering with LLMs**: The model can suggest features but needs guidance on financial relevance.
- **Modeling with LLMs**: LLMs can write and refine code, but human judgment is necessary to validate results.
- **Real-world Limitations**: Stock price prediction remains challenging; models should be used critically and not solely relied upon for trading decisions.

Next Steps:
- Experiment with different LLMs and compare outputs.
- Try refining feature engineering techniques.
- Test alternative modeling approaches, such as deep learning or ensemble methods.



# Task
Analyze historical stock price data to predict future price movements using regression and classification models, and discuss potential trading strategies based on the classification model's performance.

## Feature engineering

### Subtask:
Generate meaningful features such as moving averages, technical indicators, and volatility measures based on the loaded stock price data.


**Reasoning**:
Calculate moving averages, daily range, and daily return, then remove rows with NaN values.



In [1]:
data['MA_5'] = data['Close'].rolling(window=5).mean()
data['MA_10'] = data['Close'].rolling(window=10).mean()
data['MA_50'] = data['Close'].rolling(window=50).mean()
data['Daily_Range'] = data['High'] - data['Low']
data['Daily_Return'] = data['Close'].pct_change()
data.dropna(inplace=True)

NameError: name 'data' is not defined

**Reasoning**:
The previous code block failed because the `data` DataFrame was not defined. I need to redefine the data DataFrame by loading the data again.



In [2]:
import yfinance as yf
import pandas as pd

data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
data['MA_5'] = data['Close'].rolling(window=5).mean()
data['MA_10'] = data['Close'].rolling(window=10).mean()
data['MA_50'] = data['Close'].rolling(window=50).mean()
data['Daily_Range'] = data['High'] - data['Low']
data['Daily_Return'] = data['Close'].pct_change()
data.dropna(inplace=True)

/tmp/ipython-input-1572279735.py:4: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed


## Data preparation for modeling

### Subtask:
Split the data into training and testing sets (70/30 ratio) for both regression and classification tasks.


**Reasoning**:
Define features and target variables for regression and classification, then split the data into training and testing sets for both tasks.



In [3]:
from sklearn.model_selection import train_test_split

X = data[['Open', 'High', 'Low', 'Close', 'Volume', 'MA_5', 'MA_10', 'MA_50', 'Daily_Range', 'Daily_Return']]

y_regression = data['Close'].shift(-1).dropna()
y_classification = (data['Close'].shift(-1) > data['Close']).astype(int).dropna()

X = X.iloc[:-1]

X_train_regression, X_test_regression, y_train_regression, y_test_regression = train_test_split(
    X, y_regression, test_size=0.3, random_state=42
)

X_train_classification, X_test_classification, y_train_classification, y_test_classification = train_test_split(
    X, y_classification, test_size=0.3, random_state=42
)

print("Regression Data Shapes:")
print("X_train_regression:", X_train_regression.shape)
print("X_test_regression:", X_test_regression.shape)
print("y_train_regression:", y_train_regression.shape)
print("y_test_regression:", y_test_regression.shape)

print("\nClassification Data Shapes:")
print("X_train_classification:", X_train_classification.shape)
print("X_test_classification:", X_test_classification.shape)
print("y_train_classification:", y_train_classification.shape)
print("y_test_classification:", y_test_classification.shape)

ValueError: Found input variables with inconsistent numbers of samples: [706, 707]

**Reasoning**:
The error indicates that the number of samples in the features (X) and the classification target (y_classification) are inconsistent. This is because dropping the last row from X was done after creating y_classification. The code needs to be fixed to ensure that X and y_classification have the same number of samples before splitting.



In [4]:
from sklearn.model_selection import train_test_split

X = data[['Open', 'High', 'Low', 'Close', 'Volume', 'MA_5', 'MA_10', 'MA_50', 'Daily_Range', 'Daily_Return']]

y_regression = data['Close'].shift(-1).dropna()

# Ensure X and y_classification have the same index before creating y_classification
data['Price_Increase'] = (data['Close'].shift(-1) > data['Close']).astype(int)
y_classification = data['Price_Increase'].dropna()


# Drop the last row from X to align with the shifted target variables.
# This needs to be done after creating both y_regression and y_classification
X = X.iloc[:-1]

# Align X with the target variables after dropping the last row
y_regression = y_regression.loc[X.index]
y_classification = y_classification.loc[X.index]


X_train_regression, X_test_regression, y_train_regression, y_test_regression = train_test_split(
    X, y_regression, test_size=0.3, random_state=42
)

X_train_classification, X_test_classification, y_train_classification, y_test_classification = train_test_split(
    X, y_classification, test_size=0.3, random_state=42
)

print("Regression Data Shapes:")
print("X_train_regression:", X_train_regression.shape)
print("X_test_regression:", X_test_regression.shape)
print("y_train_regression:", y_train_regression.shape)
print("y_test_regression:", y_test_regression.shape)

print("\nClassification Data Shapes:")
print("X_train_classification:", X_train_classification.shape)
print("X_test_classification:", X_test_classification.shape)
print("y_train_classification:", y_train_classification.shape)
print("y_test_classification:", y_test_classification.shape)

Regression Data Shapes:
X_train_regression: (494, 10)
X_test_regression: (212, 10)
y_train_regression: (494, 1)
y_test_regression: (212, 1)

Classification Data Shapes:
X_train_classification: (494, 10)
X_test_classification: (212, 10)
y_train_classification: (494,)
y_test_classification: (212,)


## Model building (regression)

### Subtask:
Train a linear regression model to predict tomorrow's stock price using the engineered features.


**Reasoning**:
Train a linear regression model using the training data.



In [5]:
from sklearn.linear_model import LinearRegression

regression_model = LinearRegression()
regression_model.fit(X_train_regression, y_train_regression)

LinearRegression()

## Model evaluation (regression)

### Subtask:
Evaluate the performance of the regression model using appropriate metrics like Mean Squared Error (MSE) and R-squared (R²).


**Reasoning**:
Evaluate the performance of the trained regression model using Mean Squared Error and R-squared.



In [6]:
from sklearn.metrics import mean_squared_error, r2_score

y_pred_regression = regression_model.predict(X_test_regression)

mse = mean_squared_error(y_test_regression, y_pred_regression)
r2 = r2_score(y_test_regression, y_pred_regression)

print(f"Mean Squared Error (MSE): {mse}")
print(f"R-squared (R²): {r2}")

Mean Squared Error (MSE): 6.781462881031935
R-squared (R²): 0.9914986692076418


## Model building (classification)

### Subtask:
Create a binary target variable for price increase/decrease and train a Random Forest classifier using the engineered features.


**Reasoning**:
Import the necessary classifier and train the model.



In [7]:
from sklearn.ensemble import RandomForestClassifier

classification_model = RandomForestClassifier(random_state=42)
classification_model.fit(X_train_classification, y_train_classification)

RandomForestClassifier(random_state=42)

## Model evaluation (classification)

### Subtask:
Evaluate the performance of the classification model using metrics like accuracy, confusion matrix, and classification report. Report feature importance for the Random Forest model.


**Reasoning**:
Import necessary libraries for classification evaluation and feature importance, then perform predictions and evaluate the model. Finally, display the feature importances.



In [8]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import pandas as pd

y_pred_classification = classification_model.predict(X_test_classification)

accuracy = accuracy_score(y_test_classification, y_pred_classification)
print(f"Accuracy: {accuracy}")

conf_matrix = confusion_matrix(y_test_classification, y_pred_classification)
print(f"Confusion Matrix:\n{conf_matrix}")

class_report = classification_report(y_test_classification, y_pred_classification)
print(f"Classification Report:\n{class_report}")

feature_importances = classification_model.feature_importances_
feature_names = X_train_classification.columns

importance_df = pd.DataFrame({'Feature': feature_names, 'Importance': feature_importances})
importance_df = importance_df.sort_values(by='Importance', ascending=False)

print("\nFeature Importance:")
print(importance_df)

Accuracy: 0.5283018867924528
Confusion Matrix:
[[54 48]
 [52 58]]
Classification Report:
              precision    recall  f1-score   support

           0       0.51      0.53      0.52       102
           1       0.55      0.53      0.54       110

    accuracy                           0.53       212
   macro avg       0.53      0.53      0.53       212
weighted avg       0.53      0.53      0.53       212


Feature Importance:
            Feature  Importance
9  (Daily_Return, )    0.134432
7         (MA_50, )    0.111654
4    (Volume, AAPL)    0.111188
8   (Daily_Range, )    0.108541
6         (MA_10, )    0.093760
3     (Close, AAPL)    0.093346
0      (Open, AAPL)    0.089955
1      (High, AAPL)    0.088086
2       (Low, AAPL)    0.086025
5          (MA_5, )    0.083013


## Prediction on new data

### Subtask:
Use the trained Random Forest model to predict stock price increase/decrease for the period from 2023-01-01 to 2024-01-01.


**Reasoning**:
Download new data, engineer features, create target variable, align dataframes, and predict using the trained classification model.



In [9]:
import yfinance as yf
import pandas as pd

# 1. Download historical stock price data for Apple Inc. (AAPL) from 2023-01-01 to 2024-01-01
new_data = yf.download("AAPL", start="2023-01-01", end="2024-01-01")

# 2. Engineer the same features for new_data
new_data['MA_5'] = new_data['Close'].rolling(window=5).mean()
new_data['MA_10'] = new_data['Close'].rolling(window=10).mean()
new_data['MA_50'] = new_data['Close'].rolling(window=50).mean()
new_data['Daily_Range'] = new_data['High'] - new_data['Low']
new_data['Daily_Return'] = new_data['Close'].pct_change()

# 4. Create the target variable y_new_classification
new_data['Price_Increase'] = (new_data['Close'].shift(-1) > new_data['Close']).astype(int)

# 5. Align X_new and y_new_classification by dropping any rows with NaN values
new_data.dropna(inplace=True)

# 3. Create the features DataFrame X_new and ensure column order matches X_train_classification
feature_names = X_train_classification.columns
X_new = new_data[feature_names]

# Align y_new_classification with X_new after dropping NaNs
y_new_classification = new_data['Price_Increase']

# 6. Use the trained classification_model to predict
y_pred_new_classification = classification_model.predict(X_new)

print("Shape of X_new:", X_new.shape)
print("Shape of y_new_classification:", y_new_classification.shape)
print("Shape of y_pred_new_classification:", y_pred_new_classification.shape)

/tmp/ipython-input-1764633292.py:5: FutureWarning: YF.download() has changed argument auto_adjust default to True
  new_data = yf.download("AAPL", start="2023-01-01", end="2024-01-01")
[*********************100%***********************]  1 of 1 completed

Shape of X_new: (201, 10)
Shape of y_new_classification: (201,)
Shape of y_pred_new_classification: (201,)


## Evaluation on new data

### Subtask:
Evaluate the performance of the Random Forest model on the 2023-2024 data.


**Reasoning**:
Import the necessary metrics for model evaluation and calculate the accuracy, confusion matrix, and classification report for the predictions on the new data.



In [10]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Calculate accuracy on the new data
accuracy_new = accuracy_score(y_new_classification, y_pred_new_classification)
print(f"Accuracy on 2023-2024 data: {accuracy_new}")

# Generate confusion matrix on the new data
conf_matrix_new = confusion_matrix(y_new_classification, y_pred_new_classification)
print(f"Confusion Matrix on 2023-2024 data:\n{conf_matrix_new}")

# Print classification report on the new data
class_report_new = classification_report(y_new_classification, y_pred_new_classification)
print(f"Classification Report on 2023-2024 data:\n{class_report_new}")

Accuracy on 2023-2024 data: 0.44776119402985076
Confusion Matrix on 2023-2024 data:
[[72 20]
 [91 18]]
Classification Report on 2023-2024 data:
              precision    recall  f1-score   support

           0       0.44      0.78      0.56        92
           1       0.47      0.17      0.24       109

    accuracy                           0.45       201
   macro avg       0.46      0.47      0.40       201
weighted avg       0.46      0.45      0.39       201



## Strategy discussion

### Subtask:
Discuss potential trading strategies based on the performance of the classification model.


**Reasoning**:
Based on the previous evaluation results, I will discuss potential trading strategies, acknowledge the model's limitations, and suggest next steps for improvement.



In [11]:
print("Discussion on Potential Trading Strategies and Model Limitations:")
print("\nEvaluation on Initial Test Data (up to 2023-01-01):")
print("Accuracy:", accuracy)
print("Confusion Matrix:\n", conf_matrix)
print("Classification Report:\n", class_report)

print("\nEvaluation on New Data (2023-2024):")
print("Accuracy on 2023-2024 data:", accuracy_new)
print("Confusion Matrix on 2023-2024 data:\n", conf_matrix_new)
print("Classification Report on 2023-2024 data:\n", class_report_new)

print("\n--- Analysis ---")
print("The Random Forest classification model showed a moderate performance on the initial test data (accuracy ~52.8%), slightly better than random guessing. The classification report indicated similar performance for both classes (increase/decrease).")
print("However, when evaluated on the new data from 2023-2024, the model's performance significantly dropped (accuracy ~44.8%). The confusion matrix and classification report for the new data show a notable decrease in the model's ability to correctly predict price increases (low recall and F1-score for class 1), while it was somewhat better at identifying price decreases (higher recall for class 0, but lower precision).")

print("\n--- Potential Trading Strategies (with caution) ---")
print("Given the poor performance on recent data, relying solely on this model for trading is highly risky. However, if one were to consider strategies based on its limited insights:")
print("\nConservative Strategy: Avoid trading when the model predicts a price increase (class 1) on the new data, as its ability to correctly identify these instances is very low. Perhaps consider shorting or staying out of the market when the model predicts a decrease (class 0), although the precision for this class is not high.")
print("\nAggressive Strategy: This is not recommended given the model's performance on recent data. Any aggressive strategy based on this model would likely lead to significant losses.")

print("\n--- Limitations and Risks ---")
print("The significant drop in performance on the 2023-2024 data highlights a major limitation: the model trained on older data does not generalize well to newer market conditions. This could be due to changing market dynamics, economic factors, or other external events not captured by the engineered features.")
print("The risks of using this model for real-world trading are high, including frequent incorrect predictions, especially for price increases, leading to potential financial losses. Past performance on older data is not indicative of future results, especially with a model that doesn't generalize well.")

print("\n--- Next Steps for Improvement ---")
print("1.  **Feature Engineering:** Explore more sophisticated features, including technical indicators like RSI, MACD, Bollinger Bands, and volatility measures beyond just daily range and return. Consider incorporating economic indicators or sentiment analysis if accessible.")
print("2.  **Data Period:** Retrain the model on a more recent and potentially longer time frame that includes the data from 2023-2024 to capture current market trends.")
print("3.  **Model Selection and Tuning:** Experiment with other classification algorithms (e.g., Gradient Boosting, SVM, Neural Networks) and tune the hyperparameters of the Random Forest model.")
print("4.  **Time Series Cross-Validation:** Use time series specific cross-validation techniques to get a more robust estimate of the model's performance on future data.")
print("5.  **Incorporate External Factors:** If possible, include features related to news sentiment, social media trends, or relevant economic announcements that could influence stock prices.")

Discussion on Potential Trading Strategies and Model Limitations:

Evaluation on Initial Test Data (up to 2023-01-01):
Accuracy: 0.5283018867924528
Confusion Matrix:
 [[54 48]
 [52 58]]
Classification Report:
               precision    recall  f1-score   support

           0       0.51      0.53      0.52       102
           1       0.55      0.53      0.54       110

    accuracy                           0.53       212
   macro avg       0.53      0.53      0.53       212
weighted avg       0.53      0.53      0.53       212


Evaluation on New Data (2023-2024):
Accuracy on 2023-2024 data: 0.44776119402985076
Confusion Matrix on 2023-2024 data:
 [[72 20]
 [91 18]]
Classification Report on 2023-2024 data:
               precision    recall  f1-score   support

           0       0.44      0.78      0.56        92
           1       0.47      0.17      0.24       109

    accuracy                           0.45       201
   macro avg       0.46      0.47      0.40       201
weighted

## Summary:

### Data Analysis Key Findings

*   The linear regression model achieved a Mean Squared Error (MSE) of approximately 6.78 and an R-squared (R²) score of approximately 0.99 on the initial test data (up to 2023-01-01), indicating a strong fit to the training data.
*   The Random Forest classification model showed a moderate performance on the initial test data (up to 2023-01-01) with an accuracy of approximately 52.83%. The confusion matrix and classification report indicated similar performance for both predicting price increases and decreases.
*   Feature importance analysis for the Random Forest model on the initial data highlighted 'Daily\_Return', 'MA\_50', and 'Volume' as the most influential features.
*   When evaluated on the new data from 2023-2024, the Random Forest classification model's performance significantly decreased, with an accuracy of approximately 44.8%.
*   On the 2023-2024 data, the classification model struggled significantly with correctly predicting price increases (low recall and F1-score for class 1) but was somewhat better at identifying price decreases (higher recall for class 0, though with lower precision).

### Insights or Next Steps

*   The significant drop in the classification model's performance on recent data (2023-2024) compared to the initial test data suggests that the model trained on older data does not generalize well to newer market conditions. This highlights the dynamic nature of financial markets and the need for models to adapt to changing trends.
*   Given the poor performance on recent data, relying solely on this classification model for real-world trading is highly risky and not recommended. Future work should focus on retraining the model with more recent data, exploring more sophisticated features (including external factors), and potentially using more advanced modeling techniques suited for time series data to improve generalization and predictive power.
